### PR TITLE
Functions

### PR DESCRIPTION
**Pull request by jmdobry**

### CODE REVIEW

**Comment by jmdobry** [comment]
@jonparrott I'll stop adding to this PR so it doesn't get huge, and I'll continue with a branch of this branch.

**Comment by jonparrott** [comment]
@jmdobry are you ready for this to be reviewed?

File: functions/datastore/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<img src="https://avatars2.githubusercontent.com/u/2810941?v=3&s=96" alt="Google Cloud Platform logo" title="Google Cloud Platform" align="right" height="96" width="96"/>
```
Comment (jonparrott) [inline functions/datastore/README.md]: Is there not a tutorial for this on devsite? Would it be more valuable to house most of this documentation there?

File: functions/datastore/index.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+'use strict';
+
+var gcloud = require('gcloud');
+
+// Create a datastore client.
+var datastore = gcloud.datastore();
+
+// Gets a Datastore key from the kind/key pair in the request
+function _getKeyFromData (data, callback) {
```
Comment (jonparrott) [inline functions/datastore/index.js]: Is prefixing with an underscore the way to make things private in GCF?

Comment (jonparrott) [inline functions/datastore/index.js]: `requestData` is a better variable name than `data` (literally everything is data).

File: functions/datastore/index.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  return _getKeyFromData(data, function (err, k) {
+    if (err) {
+      return callback(err);
+    }
+
+    return datastore.get(k, function (err, entity) {
+      if (err) {
+        return callback(err);
+      }
+      return callback(null, k, entity || null);
```
Comment (jonparrott) [inline functions/datastore/index.js]: is `|| null` necessary? What does datastore return?

File: functions/datastore/index.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+// Saves (creates or inserts) an entity with the given key
+function _saveEntity (key, entity, callback) {
+  return datastore.save({
+    key: key,
+    data: entity
+  }, callback);
+}
+
+/**
```
Comment (jonparrott) [inline functions/datastore/index.js]: Why use this style of comment here but `//` elsewhere?

File: functions/gcs/index.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+var gcloud = require('gcloud');
+var readline = require('readline');
+
+module.exports = {
```
Comment (jonparrott) [inline functions/gcs/index.js]: Does it make sense separate `wordCount` into a top-level function and then do `exports.wordCount = wordCount`? Or is it accepted style to do it this way for single-function modules?

File: functions/gcs/index.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    if (!bucketName) {
+      return context.failure('Bucket not provided. Make sure you have a ' +
+        '"bucket" property in your request');
+    }
+    if (!fileName) {
+      return context.failure('Filename not provided. Make sure you have a ' +
+        '"file" property in your request');
+    }
+
+    // Create a gcs client.
```
Comment (jonparrott) [inline functions/gcs/index.js]: In a similar question, does it make sense to separate the "Cloud Storage" parts from the "Cloud Function" parts like you did in the datastore sample?

File: functions/gcs/sample.txt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+Shall I compare thee to a summer's day?
```
Comment (jonparrott) [inline functions/gcs/sample.txt]: Shakespeare is an interesting choice; I always go for Alice's Adventures in Wonderland. 😛

File: functions/log2/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<img src="https://avatars2.githubusercontent.com/u/2810941?v=3&s=96" alt="Google Cloud Platform logo" title="Google Cloud Platform" align="right" height="96" width="96"/>
```
Comment (jonparrott) [inline functions/log2/README.md]: What's different enough about these samples to necessitate two of them?

Comment (jmdobry) [inline functions/log2/README.md]: There are multiple "hello world" samples in the docs (I don't know why), which is where these samples are used. I think this particular sample will go away after the docs are re-organized.

File: functions/gcs/index.js (hunk 6 above)
Comment (jmdobry) [inline functions/gcs/index.js]: That sounds good.

File: functions/gcs/index.js (hunk 5 above)
Comment (jmdobry) [inline functions/gcs/index.js]: Will do.

File: functions/datastore/index.js (hunk 2 above)
Comment (jmdobry) [inline functions/datastore/index.js]: > Is prefixing with an underscore the way to make things private in GCF?

No, it comes from the JavaScript convention of using underscores to signal that something should be considered private.

File: functions/datastore/README.md (hunk 1 above)
Comment (jmdobry) [inline functions/datastore/README.md]: There will be a more complete tutorial added to devsite.

File: functions/datastore/index.js (hunk 3 above)
Comment (jmdobry) [inline functions/datastore/index.js]: Fixed

File: functions/datastore/index.js (hunk 4 above)
Comment (jmdobry) [inline functions/datastore/index.js]: Fixed

File: functions/gcs/index.js (hunk 6 above)
Comment (jmdobry) [inline functions/gcs/index.js]: Fixed

File: functions/gcs/index.js (hunk 5 above)
Comment (jmdobry) [inline functions/gcs/index.js]: Fixed

File: functions/pubsub/index.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+// limitations under the License.
+
+'use strict';
+
+var gcloud = require('gcloud');
+
+// Create a pubsub client.
+var pubsub = gcloud.pubsub();
+
+module.exports = {
```
Comment (jonparrott) [inline functions/pubsub/index.js]: Seems you should give this the same treatment as the other samples and define functions and assign them to `exports`.

File: functions/gcs/index.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      '"file" property in your request');
+  }
+
+  // Create a gcs client.
+  var gcs = gcloud.storage();
+  var bucket = gcs.bucket(bucketName);
+  return bucket.file(fileName).createReadStream();
+}
+
+function wordCount (context, data) {
```
Comment (jonparrott) [inline functions/gcs/index.js]: It'd be great to document what is expected of `data`.

File: functions/pubsub/index.js (hunk 9 above)
Comment (jmdobry) [inline functions/pubsub/index.js]: Done.

File: functions/gcs/index.js (hunk 10 above)
Comment (jmdobry) [inline functions/gcs/index.js]: Done.

**Comment by jonparrott** [comment]
This LGTM, once Travis is green.

**Pull request merged**